jgrapht-core/src/main/java/org/jgrapht/util/DoublyLinkedList.java

```diff
@@ -17,473 +17,1166 @@
+     * {@code reversed} is {@code true} the returned {@code NodeIterator} iterates in reverse
+     * direction returning the respective previous element in subsequent calls to
+     * {@code next(Node)}, if {@code reversed} is {@code false} it iterates in forward direction.
+     * The returned iterator ignores the actual bounds of this {@code DoublyLinkedList} and iterates
+     * until the node before the first one is reached. Its {@link NodeIterator#hasNext() hasNext()}
+     * returns {@code false} if the next node would be the first one.
+     * </p>
+     * 
+     * @param firstElement equal to the first {@code next()}
+     * @param reversed if true the returned {@code NodeIterator} iterates in reversed direction over
+     *        this list, else it iterates in forward direction.
+     * @return a wrapping {@code NodeIterator} iterating from {@code firstElement} in the specified
+     *         direction
      */
-    @Override
-    public String toString() {
-        if (isEmpty()) {
-            return "{ }";
+    public NodeIterator<E> wrappingIterator(E firstElement, boolean reversed)
+    {
+        ListNode<E> startNode = nodeOf(firstElement);
+        if (startNode == null) {
+            throw new NoSuchElementException();
         }
-        StringBuilder builder = new StringBuilder("{");
-        ListNode<E> current = head;
-        while (current.next != head) {
-            builder.append(current.value.toString()).append(", ");
-            current = current.next;
+        if (!reversed) {
+            return new ListNodeIteratorImpl(0, startNode);
+        } else {
+            return reverseIterator(new ListNodeIteratorImpl(size, startNode.next));
+        }
+    }
+
+    @Override
+    public NodeIterator<E> descendingIterator()
+    {
+        return reverseIterator(listIterator(size));
+    }
+
+    @Override
+    public NodeIterator<E> iterator()
+    {
+        return listIterator();
+    }
+
+    @Override
+    public ListNodeIterator<E> listIterator()
+    {
+        return listIterator(0);
+    }
+
+    @Override
+    public ListNodeIterator<E> listIterator(int index)
+    {
+        return new ListNodeIteratorImpl(index);
+    }
+
+    /**
+     * Returns a {@link ListNodeIterator} over the elements in this list (in proper sequence)
+     * starting with the first {@link ListNode} whose value is equal to the specified
+     * {@code element}.
+     *
+     * @param element the first element to be returned from the list iterator (by a call to the
+     *        {@code next} method)
+     * @return a list iterator over the elements in this list (in proper sequence)
+     * @throws NoSuchElementException if {@code element} is not in the list
+     */
+    public ListNodeIterator<E> listIterator(E element)
+    {
+        ModifiableInteger startIndex = new ModifiableInteger(0);
+        ListNode<E> startNode = searchNode(() -> head, n -> n.next, element, startIndex);
```

Comment:
First I thought about that solution, too. But I wanted to avoid the creation of a `Pair`- and `Integer`-Object for the cases where only the ListNode is used and the index is not used (in `nodeOf()` and `lastNodeOf()`). And this was the solution. This is also purpose of `ModifiableInteger INDEX_NOT_USED`. But this is indeed a micro optimization and simply returning a pair is more clear.
I will do that.
In this case C++ would be really handy.